Read proxy draft

vishesh opened the pull request.

vishesh committed:
ReadProxyServer: Recruiting read proxies from masterserver

vishesh committed:
readProxies: More recruitment and management stuff at random places

etschannen commented on fdbclient/FailureMonitorClient.actor.cpp:
```diff
@@ -169,7 +169,7 @@ ACTOR Future<Void> failureMonitorClientLoop(
 	TraceEvent("FailureMonitorStart").detail("IsClient", FlowTransport::transport().isClient());
 	if (FlowTransport::transport().isClient()) {
-		wait(Never());
+		// wait(Never());
```
why was this needed?

etschannen commented on fdbserver/ReadProxyServer.actor.cpp:
```diff
@@ -0,0 +1,206 @@
+
+ACTOR Future<Void> readProxyServerCore(ReadProxyInterface readProxy, Reference<AsyncVar<ServerDBInfo>> serverDBInfo) {
+	state Database cx = openDBOnServer(serverDBInfo, TaskPriority::DefaultEndpoint, true, true);
+	state ActorCollection actors(false);
```
you are not waiting on actors.getResult(), so errors thrown by reads will not kill the readProxy

etschannen commented on fdbserver/masterserver.actor.cpp:
```diff
@@ -588,20 +619,22 @@ ACTOR Future<vector<Standalone<CommitTransactionRef>>> recruitEverything( Refere
 	wait( newSeedServers( self, recruits, seedServers ) );
 	state vector<Standalone<CommitTransactionRef>> confChanges;
-	wait( newProxies( self, recruits ) && newResolvers( self, recruits ) && newTLogServers( self, recruits, oldLogSystem, &confChanges ) );
+	wait(newProxies(self, recruits) && newReadProxies(self, recruits) && newResolvers(self, recruits) &&
```
You are not monitoring read proxies for failures, run a waitFailureServer on the read proxies, and run a waitFailureClient on the master

etschannen commented on fdbserver/ReadProxyServer.actor.cpp:
```diff
@@ -0,0 +1,206 @@
+			cx->getValueCompleted->latency = timer_int() - startTime;
+			cx->getValueCompleted->log();
+
+			req.reply.send(reply.value);
```
Why send reply.value instead of reply?

vishesh committed:
ReadProxyServer: Initial skeleton implementation

- Implements a correctness failing getKey()
- Add the class types, and readProxyServer() actors.

vishesh committed:
readProxy: add waitFailureClient for read proxies and use getValue for ID

vishesh committed:
readProxy: Implement getValue

vishesh committed:
readProxy: enable getKey

vishesh committed:
readProxy: First implementation of readRange

vishesh committed:
readProxy: Refactor copying the request in getKeyValues

vishesh committed:
readProxy: Monitor masterProxiesLastChange to clientInfoLastChange

vishesh committed:
readProxies: Rename onMasterProxiesChange and related futures

Also adds DatabaseContext::onReadProxyChange() method.

vishesh committed:
readProxy: Cache a copy of ReadProxyInfo in DatabaseContext

vishesh committed:
ReadProxy: Use TransactionOption for using read proxy

vishesh committed:
ReadProxy: Support getRange()

vishesh committed:
readProxy: Throw `transaction_too_old` for the fallback case and let it be handled from client

vishesh committed:
readProxy: Refactor getRange()

vishesh committed:
Merge remote-tracking branch 'apple/master' into task/read-proxy

vishesh committed:
ReadProxy: Fix the build after master merge

vishesh committed:
ReadProxy: Change NativeAPI implementations with new types friendlier with ReadProxy

vishesh committed:
ReadProxy: Implement WatchValue and re-use NativeAPI implementation for getRange() in ReadProxy

vishesh committed:
Bump up the time lag in FastTriggeredWatch to account for ReadProxy

vishesh committed:
ReadProxy: Remove extra fields added to SSInterface

However, this adds extra hop in fallback, as the client is now
resposible for taking care of invoking the fallback path using
ReadProxy, when normal path fails in ReadProxy.

vishesh committed:
ReadProxy: GetKeyLocations from master for now in GetExactRange

Ideally, should just be able query ReadProxy directly, however
when giving ReadProxy the whole range in this case leads to normal
codepath of getRange() which fails despite the KeySelector being
resolved for reasons yet unknown. Since, keys are resolved and
getRange() figures out the shard boundaries, it shouldn't happen.

vishesh committed:
Merge remote-tracking branch 'apple/master' into task/read-proxy

vishesh committed:
ReadProxy: Some small cleanup

vishesh committed:
ReadProxy: Add to betterMasterExists()

vishesh committed:
ClusterController: Rename RoleFitnessPair to RoleFitnessTuple

As it now has three fields.

vishesh committed:
ReadProxy: Add LocalityData field to ReadProxyInterface

vishesh committed:
ReadProxy: Don't do useReadProxy=true by default

Was added for testing.
